crates/re_query_cache/src/flat_vec_deque.rs

```diff
@@ -0,0 +1,789 @@
+    assert_deque_eq(
+        &[&[20], &[21], &[22], &[1, 2, 3], &[4, 5, 6, 7], &[8, 9, 10]],
+        &v,
+    );
+
+    v.insert_with(4, FlatVecDeque::from_vecs([vec![41, 42], vec![43]]));
+    assert_deque_eq(
+        &[
+            &[20],
+            &[21],
+            &[22],
+            &[1, 2, 3],
+            &[41, 42],
+            &[43],
+            &[4, 5, 6, 7],
+            &[8, 9, 10],
+        ],
+        &v,
+    );
+
+    v.insert_with(
+        v.num_entries(),
+        FlatVecDeque::from_vecs([vec![100], vec![200, 300, 400]]),
+    );
+    assert_deque_eq(
+        &[
+            &[20],
+            &[21],
+            &[22],
+            &[1, 2, 3],
+            &[41, 42],
+            &[43],
+            &[4, 5, 6, 7],
+            &[8, 9, 10],
+            &[100],
+            &[200, 300, 400],
+        ],
+        &v,
+    );
+}
+
+// ---
+
+impl<T> FlatVecDeque<T> {
+    /// Splits the deque into two at the given index.
+    ///
+    /// Returns a newly allocated `FlatVecDeque`. `self` contains entries `[0, entry_index)`,
+    /// and the returned deque contains entries `[entry_index, num_entries)`.
+    ///
+    /// Note that the capacity of `self` does not change.
+    ///
+    /// Panics if `entry_index` is out of bounds.
+    #[inline]
+    #[must_use = "use `.truncate()` if you don't need the other half"]
+    pub fn split_off(&mut self, entry_index: usize) -> Self {
+        let value_offset = self.value_offset(entry_index);
+
+        let mut offsets = self.offsets.split_off(entry_index);
+        for offset in &mut offsets {
+            *offset -= value_offset;
+        }
+
+        Self {
+            values: self.values.split_off(value_offset),
+            offsets,
+        }
+    }
+
+    /// Shortens the deque, keeping all entries up to `entry_index` (excluded), and
+    /// dropping the rest.
+    ///
+    /// Panics if `entry_index` is out of bounds.
+    #[inline]
+    pub fn truncate(&mut self, entry_index: usize) {
+        self.offsets.truncate(entry_index);
+        self.values.truncate(self.value_offset(entry_index));
+    }
+
+    /// Removes the entry at `entry_index` from the deque.
+    ///
+    /// This is O(1) if `entry_index` corresponds to either the start or the end of the deque.
+    /// Otherwise, this requires splitting the deque into three pieces, dropping the superfluous
+    /// one, then stitching the two remaining pices back together.
+    ///
+    /// Panics if `entry_index` is out of bounds.
+    pub fn remove(&mut self, entry_index: usize) {
+        let (start_offset, end_offset) = (
+            self.value_offset(entry_index),
+            self.value_offset(entry_index + 1),
+        );
+        let offset_range = end_offset - start_offset;
+
+        if entry_index == self.num_entries() {
+            self.offsets.truncate(self.num_entries() - 1);
+            self.values.truncate(self.values.len() - offset_range);
+            return;
```

Comment:
There is an off-by-one error here. Calling `.remove(1)` on a one-element vecdeque should panic, but instead the code will remove the first entry.

Otherwise the bug only causes a missed early-out optimization.

It might be worth adding a few `#[should_panic]` tests

```suggestion
        if entry_index + 1 == self.num_entries() {
            self.offsets.truncate(self.num_entries() - 1);
            self.values.truncate(self.values.len() - offset_range);
            return;
```